### PR TITLE
Add GPU support for Linux Azure Container Instances

### PR DESCRIPTION
**Pull request by neilpeterson**
Issue - https://github.com/terraform-providers/terraform-provider-azurerm/issues/3052

### CODE REVIEW

File: website/docs/r/container_group.html.markdown [hunk 1]
```diff
@@ -134,6 +134,10 @@ A `container` block supports:
 
 * `memory` - (Required) The required memory of the containers in GB. Changing this forces a new resource to be created.
 
+~> **Note:** Gpu resources are currently only supported in Linux containers.
```
Comment (tombuildsstuff) [inline website/docs/r/container_group.html.markdown]: we tend to put `note` fields below the field they're talking about - could we move this below the line below?

File: website/docs/r/container_group.html.markdown [hunk 2]
```diff
@@ -134,6 +134,10 @@ A `container` block supports:
...
+
+* `gpu` - (Optional) Specify to deploy the container with a GPU resource.
```
Comment (tombuildsstuff) [inline website/docs/r/container_group.html.markdown]: for consistency with the other resources could we make this:

```suggestion
* `gpu` - (Optional) A `gpu` block as defined below.
```

**Comment by neilpeterson** [comment]
@tombuildsstuff thanks for the coaching and suggestions Tom. I've addressed each of them.

File: azurerm/resource_arm_container_group.go [hunk 3]
```diff
@@ -529,6 +551,20 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
 			},
 		}
 
+		if v, ok := data["gpu"]; ok {
+			gpu := v.([]interface{})
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 4]
```diff
@@ -529,6 +551,20 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
...
+			for _, v := range gpu {
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 5]
```diff
@@ -529,6 +551,20 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
...
+				gpuCount := v.(map[string]interface{})["gpu_count"]
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 6]
```diff
@@ -529,6 +551,20 @@ func expandContainerGroupContainers(d *schema.ResourceData) (*[]containerinstanc
...
+				gpuSku := v.(map[string]interface{})["gpu_sku"]
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 7]
```diff
@@ -145,6 +145,28 @@ func resourceArmContainerGroup() *schema.Resource {
...
 						},
 
+						"gpu": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							ForceNew: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"gpu_count": {
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 8]
```diff
@@ -145,6 +145,28 @@ func resourceArmContainerGroup() *schema.Resource {
...
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
+
+									"gpu_sku": {
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 9]
```diff
@@ -779,6 +815,16 @@ func flattenContainerGroupContainers(d *schema.ResourceData, containers *[]conta
...
 				}
+				if v := resourceRequests.Gpu; v != nil {
+					gpus := make([]interface{}, 0)
+					gpu := make(map[string]interface{})
+
+					gpu["gpu_count"] = *v.Count
+					gpu["gpu_sku"] = v.Sku
+					gpus = append(gpus, gpu)
+
+					containerConfig["gpu"] = gpus
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 10]
```diff
@@ -145,6 +145,28 @@ func resourceArmContainerGroup() *schema.Resource {
...
+									"gpu_count": {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
+									},
+
+									"gpu_sku": {
+										Type:     schema.TypeString,
+										Optional: true,
+										ForceNew: true,
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: azurerm/resource_arm_container_group.go [hunk 11]
```diff
@@ -145,6 +145,28 @@ func resourceArmContainerGroup() *schema.Resource {
...
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							ForceNew: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"gpu_count": {
+										Type:     schema.TypeInt,
+										Optional: true,
+										ForceNew: true,
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply azurerm/resource_arm_container_group.go]: Done, thanks.

File: website/docs/r/container_group.html.markdown [hunk 12]
```diff
@@ -184,6 +188,14 @@ A `ports` block supports:
 
 * `protocol` - (Required) The network protocol associated with port. Possible values are `TCP` & `UDP`.
 
+--
+
+A `gpu` block supports:
+
+* `gpu_count` - (Required) The number of GPUs. Allowed values are 1, 2, or 4.
+
+* `gpu_sku` - (Required) The GPU SKU. Allowed values are K80, P100, or V100.
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply website/docs/r/container_group.html.markdown]: Done, thanks.

File: website/docs/r/container_group.html.markdown [hunk 13]
```diff
@@ -184,6 +188,14 @@ A `ports` block supports:
 
 * `protocol` - (Required) The network protocol associated with port. Possible values are `TCP` & `UDP`.
 
+--
+
+A `gpu` block supports:
+
+* `gpu_count` - (Required) The number of GPUs. Allowed values are 1, 2, or 4.
```
[the remark this answers is not in the document]
Comment (neilpeterson) [reply website/docs/r/container_group.html.markdown]: Done, thanks.

File: website/docs/r/container_group.html.markdown (hunk 1 above)
Comment (neilpeterson) [reply website/docs/r/container_group.html.markdown]: Done, thanks.

File: website/docs/r/container_group.html.markdown (hunk 2 above)
Comment (neilpeterson) [reply website/docs/r/container_group.html.markdown]: Done, thanks.

File: azurerm/resource_arm_container_group_test.go [hunk 14]
```diff
@@ -211,6 +211,9 @@ func TestAccAzureRMContainerGroup_linuxComplete(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "container.0.secure_environment_variables.%", "2"),
 					resource.TestCheckResourceAttr(resourceName, "container.0.secure_environment_variables.secureFoo", "secureBar"),
 					resource.TestCheckResourceAttr(resourceName, "container.0.secure_environment_variables.secureFoo1", "secureBar1"),
+					resource.TestCheckResourceAttr(resourceName, "container.0.gpu.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "container.0.gpu.0.gpu_count", "1"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_container_group_test.go]: given the fields been updated this'll need to be:

```suggestion
					resource.TestCheckResourceAttr(resourceName, "container.0.gpu.0.count", "1"),
```

File: azurerm/resource_arm_container_group_test.go [hunk 15]
```diff
@@ -211,6 +211,9 @@ func TestAccAzureRMContainerGroup_linuxComplete(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "container.0.gpu.0.gpu_sku", "K80"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_container_group_test.go]: given the fields been updated this'll need to be:

```suggestion
					resource.TestCheckResourceAttr(resourceName, "container.0.gpu.0.sku", "K80"),
```

**Comment by tombuildsstuff** [comment]
The test suite prior to fixing the test config:

Screenshot 2019-03-17 at 19 00 24


The failing test after updating the test config:

Screenshot 2019-03-17 at 18 59 56

**Comment by neilpeterson** [comment]
@tombuildsstuff thanks a bunch for helping push this through.

**Pull request merged**